server-mux.go

```diff
@@ -153,9 +167,62 @@ func (l MuxListener) Accept() (net.Conn, error) {
 	return cmux, nil
 }
 
+func (l *MuxListener) cleanup(c net.Conn) {
+	// Block until MuxServer close receives a value then cleanup listener
+	<-l.close
+
+	if c != nil {
+		c.Close()
+	}
+
+	l.close <- l.Listener.Close()
+}
+
+// Close Listener
+func (l *MuxListener) Close() error {
+	if l == nil {
+		return nil
+	}
+
+	// send nil to unblock gofunc to close l.Listener in Accept()
+	l.close <- nil
```

Comment:
Actually, this doesn't guarantee that other go routines will be unblocked since you are receiving the 'nil' message just in the next line.